doc/L3_performance_and_scaling_enchancements_HLD.md

```diff
@@ -0,0 +1,228 @@
+
+# 2 Design
+
+## 2.1 Scaling improvements
+
+### 2.1.2 Improvement in number of ARP Entries
+
+SONiC currently supports around 2400 host entries.
+
+In our testing we found, when sending a lot of ARP/ND requests in a burst, ARP entries are getting purged from the kernel while the later set of ARP entries was still getting added.
+The sequence of add/remove is in such a way that we were never able to cross ~2400 entries .
+
+In the kernel ARP module the following attributes govern how many ARP entries are key in the Kernel ARP cache
+```
+gc_thresh1 (since Linux 2.2)
+The minimum number of entries to keep in the ARP cache. The garbage collector will not run if there are fewer than this number of entries in the cache. Defaults to 128.
+
+gc_thresh2 (since Linux 2.2)
+The soft maximum number of entries to keep in the ARP cache. The garbage collector will allow the number of entries to exceed this for 5 seconds before collection will be performed. Defaults to 512.
+
+gc_thresh3 (since Linux 2.2)
+The hard maximum number of entries to keep in the ARP cache. The garbage collector will always run if there are more than this number of entries in the cache. Defaults to 1024.
+```
+
+To increase the number of ARP/ND entries these attributes will be changed to following values for IPv4 and IPv6
+```
+net.ipv4.neigh.default.gc_thresh1=16000
+net.ipv4.neigh.default.gc_thresh2=32000
+net.ipv4.neigh.default.gc_thresh3=48000
+
+net.ipv6.neigh.default.gc_thresh1=8000
+net.ipv6.neigh.default.gc_thresh2=16000
+net.ipv6.neigh.default.gc_thresh3=32000
+```
+To increase rate of the ARP/ND packets coming to the CPU. Currently the max rate for ARP/ND is 600 packets, we will be increasing it to higher number(8000) in CoPP file  to improve the learning time.
+
+## 2.2 Performance Improvements
+
+This section elaborates the changes done to improve L3 performance in SONiC
+
+### 2.2.1 Route installation time
+
+The tables below captures the baseline route programming time for  IPv4 and IPv6 prefixes in SONiC.
+To measure route programming time, BGP routes were advertised to a SONiC router and timed how long it took for these routes to be installed in the ASIC.
+
+### Table 2: IPv4 prefix route programming time
+
+| Routes                  | time taken on AS7712(Tomahawk) |
+| ----------------------- | ------------------------------ |
+| 10k IPv4 prefix routes  | 11 seconds                     |
+| 30k IPv4 prefix routes  | 30 seconds                     |
+| 60k IPv4 prefix routes | 48 seconds                     |
+| 90k IPv4 prefix routes  | 68 seconds                     |
+
+
+
+### Table 3: IPv6 prefix route programming time
+
+| Routes                             | time taken on AS7712(Tomahawk) |
+| ---------------------------------- | -----------------------------  |
+| 10k IPv6 route with prefixes > 64b | 11 seconds                     |
+| 30k IPv6 route with prefixes >64b  | 30 seconds                     |
```

Comment:
please list out all the usecases will need perf enhancements?
Ex: 1) Sonic boot time 2) graceful restart 3) download routes etc.